### DISCUSSION TITLE
Github connection ideas and discussion

### DISCUSSION THREAD

**Discussion by SamTolmay** [lowdefy/lowdefy#474] [Ideas and feature requests]
Hi all

The purpose of the thread is to discuss use-cases and implementation ideas for a Github connection.

Github provides both a REST and GraphQL API. These are both quite extensive, so the connection would probably be a wrapper around the API that makes things like authentication a little easier.

However, we would like to know what the community would use the connection for, and how they think it should work. So we have the following questions:

- What do you want to do with a Github connection?
- How do you think a Github connection should work?

#### Current options

A use case with a simple solution is if the app needs to use files that are in the same repository as the app. They can be referenced using the `_ref` operator. The app needs to be redeployed if the files change, but if you are using something Netlify, this happens automatically. This is the approach we are using with the Lowdefy blocks documentation.

The next use case is to read file content from a public repository. This can easily be done using the raw file content URL. For example: https://raw.githubusercontent.com/lowdefy/lowdefy/main/packages/blocks/blocksAntd/src/blocks/Button/Button.json

The `AxiosHttp` connection or a GraphQl connection can also be used to interact with the API. To authenticate with the API, currently, the only option is to use [personal access tokens](https://docs.github.com/en/rest/overview/other-authentication-methods). These can be saved as a secret and included in the request headers, so they do not require anything that `AxiosHttp` does not currently support.

#### Possible future ideas 

We are investigating OAuth user authorization flows. These would allow you to redirect a user to a Github page, where the user can authorize the app to access certain APIs.

We are also thinking about a webhook feature in Lowdefy. In this case, Github could be set up to post to the webhook when certain events happen. You would then be able to work with this event in an action chain, using Lowdefy requests etc.

**Comment by astrogewgaw** [comment]
**some context**: i plan to use a lowdefy app for displaying a list of scientists working in a particular field. i cannot maintain such a huge initiative by myself, so i wish to ask the community to help out by _suggesting_ edits to the list. so the web app has to provide an interface whereby a user can do all the edits they want and submit them, but the list displayed there does not change. instead, the edits are transferred to either the same github repository as the app itself, or a another github repository that i own (and that is public). a github bot or action could then receive the data and initiate an automated pull request. this will help me moderate the edits, as i will be able to decide when to merge each of the PRs and which PRs can be rejected.

keeping all this in mind, the app should ideally be able to communicate with either the same repository it is in, or a different repository owned by the same user that owns the app's repository (through a github personal access token or something). while having the data in the same repository as the app is not that big a deal, i would ideally like to separate the data from the app itself. i am glad to see that you already have many of these use cases in mind, so i think i can rest easy :grin:!

**Comment by byang183** [comment]
Do we have any updates on this discussion? 
In our use case, we have a internal github that is using by us to host some ai project along with model config and training settings. we would like to try out using lowdefy to build an app in front of those code and data, also providing ability of manage training tasks.

**Reply by Gervwyk** [reply]
Hi @byang183 
Yeah. So with Lowdefy v4 you can easily add a custom connection and just code the function that talk to github. We've done this for our internal app to pull data from our github sprint planning board. The easiest auth approach here was just to use a personal access token. We'd like to upgrade this in the near future as we write messages back from our internal app to github issues. Here is the request function (the design is really not polished, was just "get it done" kinda thing) as an example:
```js
import getMongoDB from './lib/getMongoDB.js';
import axios from 'axios';

import query from './gql/query.js';

const transformedData = (data, project) =>
  data.map((item) => {
    const author = item.node.content?.author?.login;
    const title = item.node.content?.title;
    const url = item.node.content?.url;
    const issueNumber = item.node.content?.number;
    const closedAt = item.node.content?.closedAt;
    const createdAt = item.node.content?.createdAt;
    const bodyHTML = item.node.content?.bodyHTML;
    const issueId = item.node.content?.id;
    const updatedAt = item.node.updatedAt;
    const itemId = item.node.id;

    const repositoryNode = item.node.fieldValues.nodes.find((node) => node.repository);
    const repository = repositoryNode?.repository?.name;

    const labelsNode = item.node.fieldValues.nodes.find((node) => node.labels);
    const labels = labelsNode ? labelsNode?.labels?.nodes?.map((label) => label.name) : [];

    const pointsNode = item.node.fieldValues.nodes.find(
      (node) => node.field && node.field.name === 'Points'
    );
    const points = pointsNode?.number;

    const actualNode = item.node.fieldValues.nodes.find(
      (node) => node.field && node.field.name === 'Actual'
    );
    const actual = actualNode?.number;

    const priorityNode = item.node.fieldValues.nodes.find(
      (node) => node.field && node.field.name === 'Priority'
    );
    const priority = priorityNode?.name;

    const progressNode = item.node.fieldValues.nodes.find(
      (node) => node.field && node.field.name === 'Status'
    );
    const progress = progressNode?.name;

    const statusNode = item.node.fieldValues.nodes.find(
      (node) => node.field && node.field.name === 'Status'
    );
    const status = statusNode?.name;

    const assigneesNode = item.node.fieldValues.nodes.find(
      (node) => node.field && node.field.name === 'Assignees'
    );

    const sprintNode = item.node.fieldValues.nodes.find(
      (node) => node.field && node.field.name === 'Sprint'
    );
    const sprint = {
      duration: sprintNode?.duration,
      startDate: sprintNode?.startDate,
      title: sprintNode?.title,
    };

    const assignees = assigneesNode ? assigneesNode?.users?.nodes?.map((user) => user.login) : [];

    const cursor = item.cursor;

    return {
      _id: itemId,
      cursor,
      issue: {
        issueId,
        title,
        url,
        issueNumber,
        author,
        closedAt,
        createdAt,
        bodyHTML,
      },
      repository,
      points,
      actual,
      updatedAt,
      priority,
      progress,
      status,
      labels,
      assignees,
      sprint,
      project,
    };
  });

function filterDuplicatesByField(array, field) {
  const uniqueValues = new Set();
  return array.filter((item) => {
    if (uniqueValues.has(item[field])) {
      return false;
    } else {
      uniqueValues.add(item[field]);
      return true;
    }
  });
}

async function SyncProjectData({ request, connection }) {
  const { accessToken } = connection;

  try {
    const variables = {
      organization: 'FOO_ORG',
      count: 100,
      cursor: '',
      project: 30, // ID
    };

    const headers = {
      Authorization: `Bearer ${accessToken}`,
      'Content-Type': 'application/json',
      Accept: 'application/vnd.github+json',
    };
    let total;
    let response;
    let data = [];
    do {
      response = await axios.post(
        'https://api.github.com/graphql',
        { query, variables },
        { headers }
      );
      data = filterDuplicatesByField(
        [
          ...transformedData(response?.data?.data?.organization?.projectV2?.items?.edges ?? [], {
            id: response?.data?.data?.organization?.projectV2?.id,
            title: response?.data?.data?.organization?.projectV2?.title,
          }),
          ...data,
        ],
        '_id'
      );
      total = response.data.data.organization.projectV2.items.totalCount;
      variables.cursor = data[0].cursor;
    } while (data.length < total);

    await getMongoDB({ connection });

    await connection.mdbClient.collection('github-projects-temp').insertMany(data);
    await connection.mdbClient
      .collection('github-projects-temp')
      .aggregate([
        { $match: {} },
        {
          $merge: {
            into: 'github-projects',
            on: '_id',
            whenMatched: 'merge',
            whenNotMatched: 'insert',
          },
        },
      ])
      .toArray();
    await connection.mdbClient.collection('github-projects-temp').drop();

    return { success: true, data };
  } catch (error) {
    console.error('Error fetching GitHub team project data:', error.message);
    return [];
  }
}

SyncProjectData.schema = {};
SyncProjectData.meta = {
  checkRead: false,
  checkWrite: false,
};

export default SyncProjectData;

```
And the graphql query:
```js
const query = `query githubProjectData_v1($organization: String!, $count: Int!, $cursor: String!, $project: Int!) {
  organization(login: $organization) {
    id
    projectV2(number: $project) {
      id
      title
      items(
        last: $count
        before: $cursor
        orderBy: {direction: ASC, field: POSITION}
      ) {
        edges {
          node {
            id
            content {
              ... on Issue {
                id
                title
                url
                number
                author {
                  login
                }
                bodyHTML
                closedAt
                createdAt
              }
            }
            fieldValues(first: 10) {
              nodes {
                ... on ProjectV2ItemFieldDateValue {
                  field {
                    ... on ProjectV2Field {
                      name
                    }
                  }
                  date
                }
                ... on ProjectV2ItemFieldLabelValue {
                  labels(first: 10) {
                    nodes {
                      name
                    }
                  }
                  field {
                    ... on ProjectV2Field {
                      name
                    }
                  }
                }
                ... on ProjectV2ItemFieldNumberValue {
                  number
                  field {
                    ... on ProjectV2Field {
                      name
                    }
                  }
                }
                ... on ProjectV2ItemFieldUserValue {
                  field {
                    ... on ProjectV2Field {
                      name
                    }
                  }
                  users(first: 10) {
                    nodes {
                      name
                      login
                    }
                  }
                }
                ... on ProjectV2ItemFieldSingleSelectValue {
                  name
                  field {
                    ... on ProjectV2SingleSelectField {
                      name
                    }
                  }
                }
                ... on ProjectV2ItemFieldRepositoryValue {
                  repository {
                    name
                    url
                  }
                  field {
                    ... on ProjectV2Field {
                      name
                    }
                  }
                }
                ... on ProjectV2ItemFieldIterationValue {
                  title
                  startDate
                  duration
                  id
                  field {
                    ... on ProjectV2IterationField {
                      name
                    }
                  }
                }
              }
            }
            updatedAt
          }
          cursor
        }
        totalCount
      }
    }
  }
}`;

export default query;
```
See more on custom plugins in v4 here:
https://github.com/lowdefy/lowdefy-example-plugins

**Reply by Gervwyk** [reply]
Use the github graphql explorer to find what your query needs to be:
https://docs.github.com/en/graphql/overview/explorer

**Reply by byang183** [reply]
Cool! Thanks for the direction, I am looking into it now.
